pyro/distributions/dequantized.py

```diff
@@ -0,0 +1,82 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import math
+
+import torch
+from torch.distributions import constraints
+
+from .torch_distribution import TorchDistribution
+
+
+class DequantizedDistribution(TorchDistribution):
+    """
+    Dequantized distribution over relaxed counts.
+
+    The density of this distribution linearly interpolates the density of
+    ``base_dist``, except in the interval ``[0,1]`` where the weight of 0 is
+    doubled to account for asymmetry.
+
+    Given a distribution ``base_dist`` over nonnegative integers, this creates
+    a distribution over positive reals such that randomly rounded samples from
+    this distribution will be distributed approximately according to
+    ``base_dist``.
+
+    For example the following two models result in approximately identical
+    distributions over ``z``, conditioned on inputs ``n, p``.
+
+        # Model 1.
+        z ~ Binomial(n, p)
+
+        # Model 2.
+        r ~ DequantizedDistribution(Binomial(n, p))
+        b ~ Bernoulli(1 + floor(r) - r)      # Quantize.
+        z = floor(r) + b
+
+    Earth mover distance error is upper bounded by 1/3.
+
+    :param base_dist: A distribution with
+        ``.support == constraints.nonnegative_integer`` and
```

Comment:
I think `constraints.integer_interval` is also valid

Reply:
Most of the math is similar but we would need a separate implementation to handle `integer_interval` since `nonnegative_integer` needs special handling at zero (e.g. the `.abs()` below). I think the cleanest way to do this is to choose good names for the two distributions, e.g. `DequantizedCount` and `DequantizedInt` or something. Do you have recommendations?